src/main/java/org/truffleruby/language/methods/GetMethodObjectNode.java

```diff
@@ -0,0 +1,165 @@
+        return execute(frame, self, name, config, callerFrame);
+    }
+
+    protected abstract RubyMethod execute(Frame frame, Object self, Object name,
+            DispatchConfiguration dispatchConfig, MaterializedFrame callerFrame);
+
+    @Specialization
+    protected RubyMethod getMethodObject(
+            Frame frame,
+            Object self,
+            Object name,
+            DispatchConfiguration dispatchConfig,
+            MaterializedFrame callerFrame,
+            @CachedContext(RubyLanguage.class) RubyContext context,
+            @Cached NameToJavaStringNode nameToJavaStringNode,
+            @Cached LookupMethodOnSelfNode lookupMethodNode,
+            @Cached DispatchNode respondToMissingNode,
+            @Cached BooleanCastNode booleanCastNode,
+            @Cached ConditionProfile notFoundProfile,
+            @Cached ConditionProfile respondToMissingProfile,
+            @Cached LogicalClassNode logicalClassNode) {
+        assert this != GetMethodObjectNodeGen.getUncached() || frame == null;
+        DeclarationContext originalDeclarationContext = null;
+
+        if (frame != null) {
+            originalDeclarationContext = RubyArguments.getDeclarationContext(frame);
+
+            final DeclarationContext declarationContext = RubyArguments.getDeclarationContext(callerFrame);
+            if (declarationContext != null) {
+                RubyArguments.setDeclarationContext(frame, declarationContext);
+            }
+        }
+
+        final String normalizedName = nameToJavaStringNode.execute(name);
+        InternalMethod method = lookupMethodNode
+                .execute(frame, self, normalizedName, dispatchConfig);
+
+        if (notFoundProfile.profile(method == null)) {
+            final Object respondToMissing = respondToMissingNode
+                    .call(self, "respond_to_missing?", name, dispatchConfig.ignoreVisibility);
+            if (respondToMissingProfile.profile(booleanCastNode.executeToBoolean(respondToMissing))) {
+                /** refinements should not affect BasicObject#method_missing */
+                RubyArguments.setDeclarationContext(frame, originalDeclarationContext);
+                final InternalMethod methodMissing = lookupMethodNode
+                        .execute(frame, self, "method_missing", dispatchConfig);
+                method = createMissingMethod(self, name, normalizedName, methodMissing, context);
+            } else {
+                throw new RaiseException(
+                        context,
+                        context.getCoreExceptions().nameErrorUndefinedMethod(
+                                normalizedName,
+                                logicalClassNode.execute(self),
+                                this));
+            }
+        }
+        final RubyMethod instance = new RubyMethod(
+                context.getCoreLibrary().methodClass,
+                RubyLanguage.getCurrentLanguage().methodShape,
```

Comment:
`RubyLanguage.getCurrentLanguage()` is slow, it's essentially a ThreadLocal lookup.
So I switched to `@CachedLanguage` instead.